### ISSUE TITLE
Grape logger always log 500 http response code for rescued exceptions

### ISSUE THREAD

**Issue by nbulaj** [ridiculous/grape-middleware-logger#10]
Hi. While exploring the `grape-middleware-logger` logs I discovered that all requests with errors that are preprocessed with `rescue_from` are marked with 500 response code:

```ruby
# api.rb
class API < Grape::API
  use Grape::Middleware::Logger

  format :json

  include ExceptionHandler

  mount V1::Base
end

# exception_handler.rb
module ExceptionHandler
  extend ActiveSupport::Concern

  included do
    rescue_from ActiveRecord::RecordNotFound do
      error!({ error: :not_found }, 404)
    end

    # ... other :rescue_from ...
  end
end
```

Log:

> I, [2016-09-14T14:53:49.896641 #27381]  INFO -- : [f4b3b020-ef4b-44ea-9228-fc58d35c1687]   ActiveRecord::RecordNotFound (Couldn't find User with 'id'=1010):

I looked at the source code of the gem and added some trace to the code. And what am I found is that `rescue_from` are invoked after  `grape-middleware-logger` methods such as `after_exception` and `after_failure`. So the real status code (and maybe some extra info) does not fall into the log, only 500. I think that is a wrong behaviour.

Gemfile.lock

> grape (0.17.0)
> grape-middleware-logger (1.7.0)

What do you think about that, @ridiculous?

**Comment by ridiculous** [comment]
Yeah, that's not ideal behavior. But unfortunately I don't see how to hook into the error handlers registered with `rescue_from` and relay those responses to the log. I can dig around some more, and will report back if I find anything

**Comment by nbulaj** [comment]
@ridiculous we can get all the rescue handlers:

```ruby
rescue_handlers = @env['api.endpoint'].namespace_reverse_stackable(:rescue_handlers)
```
And found an executable proc for the catched exception by it's class:

```ruby
{ActiveRecord::RecordInvalid=>#<Proc:0x00000005ce4960@/home/bulaj/projects/some_name/app/api.rb:14>}
```

We can invoked a proc and get its results. Usually there is a `error!` method present in the `rescue_from` block that returns `Rack::Response` from where we can get status and other info.

I'm still thinking about implementation...

**Comment by ridiculous** [comment]
@nbulaj Check out this implementation https://github.com/ridiculous/grape-middleware-logger/pull/15

It's the most efficient way I see to do it. The main issue is that the `:rescue_handlers` are processed by `Grape::Middleware::Error`, and there's no way to hook into that

**Comment by ridiculous** [comment]
I didn't bother to update the tests yet. Let me know what you think and we go from there

**Comment by nbulaj** [comment]
Hi @ridiculous. Any news about that issue? :)

**Comment by ridiculous** [comment]
Nothing new. What do you think of my solution in #15, overloading `Rack::Head#call`? It's the only solution I can think of, even though it's hacky. The problem of not being able to get the error + code, because it's handled by a default Grape middleware and can't be rearranged, is a fundamental flaw in Grape's stack design.

**Comment by ridiculous** [comment]
If you're fine with #15, I'm happy to release a new version with it

**Comment by dakull** [comment]
same issue here - how about a simpler version with no overloading but a configurable status code per exception class? e.g. validation errors would get a 400/422 and everything else a 500.

**Comment by nbulaj** [comment]
@dakull but Grape [already can have custom exception statuses](https://github.com/ruby-grape/grape#exception-handling) and it will be some kind of overhead (to set the status code multiple times). I think that overloading of `Rack::Head` is not a problem - look at Roda plugin system, it's doing the same thing

**Comment by pkmiec** [comment]
I'm running in this issue as well. Why not,
```
insert_before ::Grape::Middleware::Error, ::Grape::Middleware::Logger, condensed: true
```
???

**Comment by ridiculous** [comment]
Did you try it? Does it work?

**Comment by pkmiec** [comment]
Yes, I tried it and it works :). However, since I just started using grape recently, I don't have enough experience to say whether there are issues with this approach. So, it would be good for someone with more grape experience / understanding to try it as well.
